### PR TITLE
conservative approach in Observable to InputStream adapter + Unit test

### PR DESCRIPTION
**Pull request by maksimlikharev**
1. Unit test to support the change, this fails previous implementation every time
2. New Observable to InputStream adapter
3. HttpContentInputStream changed to ObservableInputStreamAdapter

### CODE REVIEW

File: karyon2-jersey-blocking/src/main/java/netflix/karyon/jersey/blocking/ObservableInputStreamAdapter.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import rx.Observable;
+import rx.Observer;
+
+public class ObservableInputStreamAdapter extends InputStream {
```
Comment (NiteshKant) [inline karyon2-jersey-blocking/src/main/java/netflix/karyon/jersey/blocking/ObservableInputStreamAdapter.java]: Any reason why this fix can not be made to the existing HttpContentInputStream?

**Comment by NiteshKant** [comment]
Thanks @maksimlikharev for the PR.

It will help me understand your approach if you can explain the problem with the existing `HttpContentInputStream` that you faced. I think this is for the issue #198 you created but it does not contain enough details about the issue, specifically around which part of the `HttpContentInputStream` code is causing loss of data.

**Comment by maksimlikharev** [comment]
Reason was to have both cases for comparison, also would result almost override of the existing file.
In case you prefer change  done in HttpContentInputStream, I can do it

**Comment by maksimlikharev** [comment]
in essence

````
    public int available() throws IOException {
        return isCompleted ? contentBuffer.readableBytes() : 0;
    }
````

when not complete, it will return 0

````
	public int read() throws IOException {
		int available = available();
		if (available == 0) {
			return -1;
		}
````

in here it reads, if available == 0, it will indicate end of data.

Another reason for the new class, on my taste this class complicates things beyond necessity, some places are synchronized and blocked, some are not. This is why I used more conservative approach.

**Comment by NiteshKant** [comment]
>> in here it reads, if available == 0, it will indicate end of data.

Isn't it so that `ReadFunc.read()` is only called when the content stream is completed? (from [here](https://github.com/Netflix/karyon/blob/master/karyon2-governator/src/main/java/netflix/karyon/transport/util/HttpContentInputStream.java#L122))

>> In case you prefer change done in HttpContentInputStream, I can do it

I will prefer either to remove `HttpContentInputStream` or do the changes there. I see that your implementation does not support `mark`, any reason why?

>> Another reason for the new class, on my taste this class complicates things beyond necessity, some places are synchronized and blocked, some are not.

The code only synchronizes to access the monitor `contentAvailabilityMonitor` which I think is the only place requiring synchronization. I agree that the code has a bug where in it does not signal the monitor on termination of the content stream, which will cause issues for empty payloads. I see that your implementation does notify on termination, so :+1: :)

**Comment by maksimlikharev** [comment]
Nitesh,
I didn't do mark as stock InputStream implementation not supporting that as well, same Ockham principal, but can be added there, as we are really buffer here. 
You the owner, so please decide to replace or fix :)
I would really prefer to reference Netflix artifact as long as it works. 

Please let me know if I have to do anything here.

**Comment by NiteshKant** [comment]
> Please let me know if I have to do anything here.

I am more than happy to accept this PR, I am just making sure that I understand the issue. Can you answer my question:

 >> in here it reads, if available == 0, it will indicate end of data.

>>>> Isn't it so that `ReadFunc.read()` is only called when the content stream is completed? (from [here](https://github.com/Netflix/karyon/blob/master/karyon2-governator/src/main/java/netflix/karyon/transport/util/HttpContentInputStream.java#L122))

**Comment by maksimlikharev** [comment]
This is my view on the problem

If completed, go here, no problem here, this is volatile

```
        if (isCompleted) {
            if (!contentBuffer.isReadable()) {
                return -1;
            } else {
                return readFunc.read();
            }
```

if nothing to read, go here and possibly block, please note __!contentBuffer.isReadable()__, this is not synchronized or volatile, phantom/dirty reads, so it falls into the else and read.

```
        } else if (!contentBuffer.isReadable()) {
            synchronized (contentAvailabilityMonitor) {
                try {
                    contentAvailabilityMonitor.wait();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt(); // Reset the interrupted flag for the upstream to handle.
                    logger.error("Interrupted while waiting for content.", e);
                    throw new IOException("Interrupted while waiting for content.", e);
                }
            }

            /*Do not call within the sync method.*/ return read(); // Since we have been notified, we check again to see if content is available.
        } else {
            return readFunc.read();
        }

**Comment by NiteshKant** [comment]
@maksimlikharev Apologies for not replying earlier. I understand the issue now w.r.t the race-condition. I propose doing the following changes:

- Make the changes you have here in `ObservableInputStreamAdapter` to `HttpContentInputStream` 
- Preserve the `mark`, `reset`, `skip` behavior.
- Include `contentBuffer` release in `close()` of `InputStream`
- I see that this PR has some more commits unrelated to this change, remove those commits (move them to a different PR)

WDYT?

**Comment by maksimlikharev** [comment]
Ok, I'll do this on Wednesday

**Comment by NiteshKant** [comment]
Thanks @maksimlikharev !

**Comment by maksimlikharev** [comment]
Nitesh,
I produce all necessary changes and removed changes from Admin site as well, but now Jenkins cannot merge it automatically ( Admin site ). What needs to be done here?

**Comment by maksimlikharev** [comment]
Ok, I merged from upstream, resolved conflict, now I'm afraid that PR will be huge, I need word of advice here. Thanks.

**Comment by NiteshKant** [comment]
> Ok, I merged from upstream, resolved conflict, now I'm afraid that PR will be huge, I need word of advice here.

Looks like the PR will create conflicts and thats why it is not mergeable.

May be an easier way would be for you to create a branch from master and cherry pick your changes from this PR on to that branch. You can later on hard reset your changes on your fork's master.

**Comment by maksimlikharev** [comment]
things got messy, closing this one, will refork and produce new request

**Comment by maksimlikharev** [comment]
new PR created, only the relevant changes

**Pull request closed without merging**